### PR TITLE
TGUI Outfit Manager and Editor

### PR DESCRIPTION
**Pull request by TheTrigg**
## About The Pull Request

The Outfit Manager has been replaced with a pretty TGUI interface.

Custom Outfits are now loaded and saved using SSpersistence.

I never really messed around with events and stuff when I was part of a downstream server, so it's hard for me to figure out what would be useful and what would just be too situational; so feedback is always useful.

## Why It's Good For The Game

Outfits for events can be created and tweaked much more easily, and with pretty UIs in the process.

https://user-images.githubusercontent.com/36010999/113460712-0d546180-93f0-11eb-80bd-aef0a0c58e63.mp4

https://user-images.githubusercontent.com/36010999/114125551-169c6d00-98cd-11eb-8dbd-f959a546f7f3.mp4

## Changelog
:cl: Trigg
admin: Admins just got a new TGUI Outfit Manager to mess around and make events with!
admin: Custom Outfits are now persistent and saved server-side, but can still be locally saved and loaded.
/:cl:

### CODE REVIEW

File: code/modules/admin/outfit_editor.dm [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+/datum/outfit_editor/ui_status(mob/user, datum/ui_state/state)
+	if(QDELETED(drip))
+		return UI_CLOSE
+	return ..()
+
+/datum/outfit_editor/ui_close(mob/user)
+	clear_human_dummy(dummy_key)
+	qdel(src)
+
+/datum/outfit_editor/proc/init_dummy()
```
Comment (TriggeredBoi) [inline code/modules/admin/outfit_editor.dm:38]: This is the second time I've used dummies in pretty much the same way, maybe I should make some centralized way of using dummy lookalikes. Not in this PR though.
By the way I haven't forgotten about the `equip` problem pointed out in https://github.com/tgstation/tgstation/issues/57865

File: code/modules/admin/outfits.dm [hunk 2]
```diff
@@ -1,35 +1,14 @@
...
-		var/datum/outfit/varedit/VO = O
-		if(istype(VO))
-			vv = length(VO.vv_values)
-		dat += "<li>[O.name][vv ? "(VV)" : ""]</li> <a href='?_src_=holder;[HrefToken()];save_outfit=1;chosen_outfit=[REF(O)]'>Save</a> <a href='?_src_=holder;[HrefToken()];delete_outfit=1;chosen_outfit=[REF(O)]'>Delete</a>"
-	dat += "</ul>"
-	dat += "<a href='?_src_=holder;[HrefToken()];create_outfit_menu=1'>Create</a><br>"
-	dat += "<a href='?_src_=holder;[HrefToken()];load_outfit=1'>Load from file</a>"
-	admin << browse(dat.Join(),"window=outfitmanager")
 
 /datum/admins/proc/save_outfit(mob/admin,datum/outfit/O)
```
Comment (TriggeredBoi) [inline code/modules/admin/outfits.dm:4]: Why are these procs under `/datum/admins`? Can/should I move them?

File: code/datums/outfit.dm [hunk 3]
```diff
@@ -347,6 +347,35 @@
...
+	OUTFIT_COPY(l_pocket)
+	OUTFIT_COPY(r_pocket)
+	OUTFIT_COPY(suit_store)
+	OUTFIT_COPY(r_hand)
+	OUTFIT_COPY(l_hand)
+	OUTFIT_COPY(internals_slot)
+	OUTFIT_COPY(backpack_contents)
+	OUTFIT_COPY(box)
+	OUTFIT_COPY(implants)
+	OUTFIT_COPY(accessory)
```
Comment (Kylerace) [inline code/datums/outfit.dm:376]: not really a need to have this define, it doesnt save on much typing and is arguably less clear

**Comment by TriggeredBoi** [comment]
Ah fudge I forgot to make it a draft sorry

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
+import { useBackend } from '../backend';
+import { Button, Section, Box, Stack } from '../components';
+import { Window } from '../layouts';
+
+export const OutfitEditor = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { outfit, saveable } = data;
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:7]: dummy64 should be in this data, in a few spots i see you using data.thing but that's bad for readability

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+  return (
+    <Window
+      width={380}
+      height={625}>
+      <Window.Content>
+
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:14]: remove this space

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+          py={3}
+          position="absolute"
+          as="img"
+          src={`data:image/jpeg;base64,${data.dummy64}`}
+          height="100%"
+          width="100%"
+          style={{
+            '-ms-interpolation-mode': 'nearest-neighbor',
+          }} />
+
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:26]: remove this space

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            </>
+          }
+          buttons={
+            <>
+              <Button
+                icon="code"
+                tooltip="Edit this outfit on a VV window"
+                tooltipPosition="left"
+                onClick={() => act("vv", {})}
+              />
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:47]: move this to the line above

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+              <Button
+                icon="save"
+                disabled={!saveable}
+                tooltip={!!saveable && "Save this outfit to the custom outfit list"}
+                tooltipPosition="left"
+                title={!saveable && "This outfit is already on the custom outfit list. Any changes made here will be immediately applied."}
+                onClick={() => act("save", {})}
+              />
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:55]: move this to the line above

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+            </>
+          }>
+
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:58]: remove this space

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            </Stack>
+            <Stack mb={2}>
+              <OutfitSlot name="Shoes" icon="socks" slot="shoes" />
+              <OutfitSlot name="Left Pocket" slot="l_pocket" />
+              <OutfitSlot name="Right Pocket" slot="r_pocket" />
+            </Stack>
+          </Box>
+
+        </Section>
+
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:92]: two needless spaces here

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  return (
+    <Stack.Item grow={1} basis={0}>
+      <Button fluid height={2}
+        icon={icon}
+        title="Ctrl-Click to select any item"
+        onClick={e => {
+          e.ctrlKey
+            ?act("ctrlClick", { slot: slot })
+            :act("click", { slot: slot }); }}
+      >
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:111]: move this to the line above, also cool use of ctrlKey

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+      </Window.Content>
+    </Window>
+  );
+};
+
+const OutfitSlot = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { name, icon, slot } = props;
+  const currItem = data.outfit[slot];
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:101]: this should just be in a { } = data

File: tgui/packages/tgui/interfaces/OutfitManager.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+export const OutfitManager = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { outfits } = data;
+
+  return (
+    <Window
+      width={300}
+      height={300}>
+      <Window.Content>
+
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitManager.js:14]: remove this space

File: tgui/packages/tgui/interfaces/OutfitManager.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+                      tooltip="Delete this outfit"
+                      tooltipPosition="left"
+                      onClick={() => act("delete", { outfit: outfit.ref })} />
+                  </Stack.Item>
+                </Stack>
+              </Stack.Item>
+            ))}
+          </Stack>
+        </Section>
+
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitManager.js:85]: remove this space

File: tgui/packages/tgui/interfaces/OutfitEditor.js [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            <Stack mb={2}>
+              <OutfitSlot name="Belt" slot="belt" />
+              <OutfitSlot name="Left Hand" slot="l_hand" />
+              <OutfitSlot name="Right Hand" slot="r_hand" />
+            </Stack>
+            <Stack mb={2}>
+              <OutfitSlot name="Shoes" icon="socks" slot="shoes" />
+              <OutfitSlot name="Left Pocket" slot="l_pocket" />
+              <OutfitSlot name="Right Pocket" slot="r_pocket" />
+            </Stack>
```
Comment (tralezab) [inline tgui/packages/tgui/interfaces/OutfitEditor.js:88]: this code above here is pretty dope through and through, i hope you give each slot an icon like you did with the shoes

**Review by tralezab — COMMENTED** [review]
here's what i would change (or just a few compliments on how you did bits and bobs, all in all this ui is super innovative and really cool

File: tgui/packages/tgui/interfaces/OutfitEditor.js (hunk 15 above)
Comment (TriggeredBoi) [reply tgui/packages/tgui/interfaces/OutfitEditor.js:88]: Thanks - and absolutely, that's the plan.

**Comment by Pumpkinoe** [comment]
this is so cool

**Comment by AnturK** [comment]
Instead of jsons you could just dump these into local storage now. Would need changes to validation.

**Comment by TriggeredBoi** [comment]
> Instead of jsons you could just dump these into local storage now. Would need changes to validation.

uhhhhhhh how?

**Comment by AnturK** [comment]
Sorry for late reply, you can take a look at [pod launcher interface](https://github.com/tgstation/tgstation/blob/master/tgui/packages/tgui/interfaces/CentcomPodLauncher.js#L760) for quick example. Mind you'll have to validate if paths stored in the storage are still up to date when using them again.

**Comment by TriggeredBoi** [comment]
> Sorry for late reply, you can take a look at [pod launcher interface](https://github.com/tgstation/tgstation/blob/master/tgui/packages/tgui/interfaces/CentcomPodLauncher.js#L760) for quick example. Mind you'll have to validate if paths stored in the storage are still up to date when using them again.

After banging my head against that example and finally figuring out how it worked, I don't believe `storage` is the best option for storing outfits because:
1 - Outfits would not be shared between admins (but I still really like the idea of them carrying over between servers)
2 - While I can see how I could maybe do it, it wouldn't be the prettiest of implementations as I pack more than just the typepath of each slot to the OutfitEditor [in ui_data](https://github.com/tgstation/tgstation/pull/58125/files#diff-01286cdbd93c2a09be2c0549fb6f7de9392c0cd78a4bf2dcfe3236a5cf18d654R55)

As an alternative, however, I plan on figuring out how the DB stuff works tomorrow and maybe store them there instead... Or would that be too overkill?

**Comment by TriggeredBoi** [comment]
Oh hey, it worked. Even though it took me the entire day to do it, it was still easier than I thought it would be. Cool.
So GLOB.outfits is saved to a json on roundend now, and outfits with duplicate names are still allowed.

I still have a couple loose ends to fix up, after that this PR will probably be ready for review (I already feel bad for poor man who gets to review my still trash DM code)

**Comment by TriggeredBoi** [comment]
It's time. I feel bad for the reviewer, something tells me there will be a lot to address.

**Review by stylemistake — APPROVED** [review]

**Comment by TriggeredBoi** [comment]
...Or not

**Pull request merged**